### PR TITLE
Skins: allow custom brightness threshold for hotcue colors

### PR DESCRIPTION
**Pull request by ronso0**
The style of WHotcueButton can be adjusted for dim & bright hotcue colors.
In LateNight PaleMoon theme which is rather dim the default brightness threshold `127` doesn't work well; colors are considered 'dark' too early.

This change allows to set a custom `<DimBrightThreshold>` (int 0-256) for the following widgets:
* HotcueButton
* Overview
* Visual (scrolling waveforms)

In the PaleMoon example I used the default 127 in the overview to keep the contrasting marker strokes.
(see also #3003)

### CODE REVIEW

**Comment by ronso0** [comment]
Now waveforms and overviews can read the `<DimBrightThreshold>` node, too. Default is 127.

The sole purpose is to adjust cue colors to dim PaleMoon theme. Please test!

File: src/waveform/renderers/waveformsignalcolors.h [hunk 1]
```diff
@@ -25,6 +25,9 @@ class WaveformSignalColors {
     inline const QColor& getPlayedOverlayColor() const { return m_playedOverlayColor; }
     inline const QColor& getPassthroughOverlayColor() const { return m_passthroughOverlayColor; }
     inline const QColor& getBgColor() const { return m_bgColor; }
+    inline const int& getDimBrightThreshold() const {
```
Comment (Holzhaus) [inline src/waveform/renderers/waveformsignalcolors.h:28]: Don't use a const reference, just return int.

File: src/widget/wwaveformviewer.cpp [hunk 2]
```diff
@@ -27,7 +27,8 @@ WWaveformViewer::WWaveformViewer(
           m_bScratching(false),
           m_bBending(false),
           m_pCueMenuPopup(make_parented<WCueMenuPopup>(pConfig, this)),
-          m_waveformWidget(nullptr) {
+          m_waveformWidget(nullptr),
+          m_dimBrighTreshold(127) {
```
Comment (Holzhaus) [inline src/widget/wwaveformviewer.cpp:31]:
```suggestion
          m_dimBrightTreshold(127) {
```
and everywhere else

**Review by Holzhaus — CHANGES_REQUESTED** [review]
Thanks. I can't test properly, because it hangs very long when starting and loading a track doesn't work either, but I doubt this was caused by your changes. Probably this is because this branch is so old. I suggest you rebase on latest 2.3 and also change the target branch to 2.3.

File: src/widget/wwaveformviewer.cpp (hunk 2 above)
Comment (ronso0) [reply src/widget/wwaveformviewer.cpp:31]: nice catch.

File: src/waveform/renderers/waveformsignalcolors.h (hunk 1 above)
Comment (ronso0) [reply src/waveform/renderers/waveformsignalcolors.h:28]: just `inline int getDimBrightThreshold() const {` then?

Comment (Holzhaus) [reply src/waveform/renderers/waveformsignalcolors.h:28]: Yes.

**Comment by ronso0** [comment]
> Thanks. I can't test properly, because it hangs very long when starting and loading a track doesn't work either, but I doubt this was caused by your changes. Probably this is because this branch is so old. I suggest you rebase on latest 2.3 and also change the target branch to 2.3.

This is based on 2.3 already and FWIW I didn't encounter the issues you did.
I rebased and targeted at 2.3

**Comment by Holzhaus** [comment]
> > Thanks. I can't test properly, because it hangs very long when starting and loading a track doesn't work either, but I doubt this was caused by your changes. Probably this is because this branch is so old. I suggest you rebase on latest 2.3 and also change the target branch to 2.3.
> 
> This is based on 2.3 already and FWIW I didn't encounter the issues you did.
> I rebased and targeted at 2.3

Yes, I got some weird ACPI issues on my new laptop. A reboot fixed them.

This is how it looks on my machine. I think we need light hotcue button labels, but other than it it looks fine.

Screenshot from 2020-08-17 23-01-13

**Comment by ronso0** [comment]
...and again the SVGs are rendered diferently it seems:
image

now that it works I can increase the blur, but I'm a bit tired of this cross-computer testing by now, so I may as well switch to bright icons, even if that's what I wanted to avoid in the first place.

Thanks for testing!

**Comment by Holzhaus** [comment]
> ...and again the SVGs are rendered diferently it seems:
> image
> 
> now that it works I can increase the blur, but I'm a bit tired of this cross-computer testing by now, so I may as well switch to bright icons, even if that's what I wanted to avoid in the first place.

Whoops, the low contrast might actually not be Qt's fault. It's already late and I forgot that I have redshift enabled on my computer, so this might have caused this. Anyway, I think the contrast is a bit too low especially when you turn down screen brightness in a club setting, so using white labels or conderably more blur would probably be a good idea. I'd probably prefer white text instead of a strong blue effect.

**Comment by ronso0** [comment]
Okay, good to know :)
I'll play around with a bit, maybe bright icons are more cnsistent in the end considering the waveform labels also switch to bright fonts. I'm just glad it doesn't pick a bright text for the orange in that picture, that was really a hard contrast at night.

File: src/waveform/renderers/waveformsignalcolors.cpp [hunk 3]
```diff
@@ -76,6 +76,7 @@ bool WaveformSignalColors::setup(const QDomNode &node, const SkinContext& contex
     }
     m_bgColor = WSkinColor::getCorrectColor(m_bgColor).toRgb();
 
+    m_dimBrightThreshold = context.selectInt(node, QStringLiteral("DimBrightThreshold"));
```
Comment (daschuer) [inline src/waveform/renderers/waveformsignalcolors.cpp:79]: I think we should do here something like this to have the value optional: 
```  
    bool okay = false;
    m_dimBrightThreshold = context.selectInt(node, QStringLiteral("DimBrightThreshold"), &okay);
    if (okay) {
        m_dimBrightThreshold = kDefaulDimBrightThreshold;
    }
```

File: src/waveform/renderers/waveformsignalcolors.h [hunk 4]
```diff
@@ -45,6 +48,7 @@ class WaveformSignalColors {
     QColor m_playedOverlayColor;
     QColor m_passthroughOverlayColor;
     QColor m_bgColor;
+    int m_dimBrightThreshold = 127;
```
Comment (daschuer) [inline src/waveform/renderers/waveformsignalcolors.h:51]: We avoid initialize value in the class definition, because this introduces a hidden static constructor.  
Please move this to the constructor and use kDefaulDimBrightThreshold.

File: src/waveform/renderers/waveformwidgetrenderer.cpp [hunk 5]
```diff
@@ -262,6 +263,8 @@ void WaveformWidgetRenderer::setup(
         m_orientation = Qt::Horizontal;
     }
 
+    m_dimBrightTreshold = context.selectInt(node, QStringLiteral("DimBrightThreshold"));
```
Comment (daschuer) [inline src/waveform/renderers/waveformwidgetrenderer.cpp:266]: The same here. Please default to 127

File: src/widget/whotcuebutton.cpp [hunk 6]
```diff
@@ -12,9 +12,10 @@ WHotcueButton::WHotcueButton(const QString& group, QWidget* pParent)
           m_hotcue(Cue::kNoHotCue),
           m_hoverCueColor(false),
           m_pCoColor(nullptr),
-          m_cueColorDimmed(false),
-          m_isCueColorLight(false),
-          m_isCueColorDark(false) {
+          m_cueColorDimThreshold(127),
```
Comment (daschuer) [inline src/widget/whotcuebutton.cpp:15]: This should become a const in the anonymous namespace

File: src/widget/whotcuebutton.cpp [hunk 7]
```diff
@@ -28,6 +29,7 @@ void WHotcueButton::setup(const QDomNode& node, const SkinContext& context) {
     } else {
         SKIN_WARNING(node, context) << "Hotcue value invalid";
     }
+    m_cueColorDimThreshold = context.selectInt(node, QStringLiteral("DimBrightThreshold"));
```
Comment (daschuer) [inline src/widget/whotcuebutton.cpp:32]: This should not keep the default if the value is not defined in skins.

File: src/widget/whotcuebutton.cpp [hunk 8]
```diff
@@ -132,13 +134,13 @@ void WHotcueButton::slotColorChanged(double color) {
...
-        m_isCueColorLight = !m_cueColorDimmed;
-        m_isCueColorDark = m_cueColorDimmed;
+        m_bCueColorIsLight = !m_bCueColorDimmed;
+        m_bCueColorIsDark = m_bCueColorDimmed;
     } else {
         // We are now at the background set by qss.
         // Since we don't know the color reset both
-        m_isCueColorLight = false;
-        m_isCueColorDark = false;
+        m_bCueColorIsLight = false;
```
Comment (daschuer) [inline src/widget/whotcuebutton.cpp:142]: In Mixxx we have no prefix for bools. The old version was IMHO fine.

**Review by daschuer — CHANGES_REQUESTED** [review]
Thank you for this useful PR. I have left some comments.

**Comment by ronso0** [comment]
Just to let you know: I addressed the comments but didn't push, yet, because I need some continous time to thoroughly test the thresholds with updated hotcue icons and the waveforms.
This won't happen before next week, though.

File: src/widget/whotcuebutton.cpp (hunk 7 above)
Comment (ronso0) [reply src/widget/whotcuebutton.cpp]: done.

**Comment by ronso0** [comment]
A TMP commit slipped in so I finally cleaned up and rebased onto 2.3. Please have a look!

I adjusted the thresholds in LateNight PaleMoon and made the hotcue button icons more clear.
The comparison uses the new icons for demonstration.
before (general threshold = 127):
Screenshot_2020-09-23_15-27-19

this PR (custom threshold for buttons and waveform = 90):
Screenshot_2020-09-23_15-27-19___90

**Comment by Holzhaus** [comment]
90 is okay, 80 makes the "6" barely readable.

**Review by Holzhaus — COMMENTED** [review]
There are merge conflicts.

Also, I think the contrast for the blue hotcue colors from the default palette is bad. And maybe the red and purple button should be considered "dark", too?

Screenshot from 2020-10-18 16-31-07

**Comment by ronso0** [comment]
you can try a different threshold, set in skin.xml in PaleMoon color scheme

**Comment by Holzhaus** [comment]
> There are merge conflicts.
> 
> Also, I think the contrast for the blue hotcue colors from the default palette is bad. And maybe the red and purple button should be considered "dark", too?
> 
> Screenshot from 2020-10-18 16-31-07

I tried, using the light grey text on red make the contrast even worse.

**Comment by ronso0** [comment]
I'll check if it can be fixed with a brighter font for dark buttons.

**Comment by ronso0** [comment]
Oh, the first time for quite a while I got a "Build passed" message from our friend Travis :)

**Comment by ronso0** [comment]
> I think the contrast for the blue hotcue colors from the default palette is bad

I made the 'dark' icons a bit brighter:
image

> And maybe the red and purple button should be considered "dark", too?

I don't want to tweak our dark/light algorythm here. Would you mind fiing a bug report?

**Comment by Holzhaus** [comment]
> > And maybe the red and purple button should be considered "dark", too?
> 
> I don't want to tweak our dark/light algorythm here. Would you mind fiing a bug report?

The algorithm is okay, I tweaked the threshold to `105` and now it looks like this:

Screenshot from 2020-10-21 11-55-00

I think that has better contrast, but you decide. Just let me know, so we can merge this.

**Review by daschuer — APPROVED** [review]
Thank you.

**Comment by ronso0** [comment]
> > > And maybe the red and purple button should be considered "dark", too?
> > 
> > 
> > I don't want to tweak our dark/light algorythm here. Would you mind fiing a bug report?
> 
> The algorithm is okay, I tweaked the threshold to `105` and now it looks like this:

I get the impression what I'm trying to improve here is not matter of the dim/bright threshold, but it also seems to depend a lot on the screen used and _maybe also_ the algorhythm, because when I tested with Mixxx Hotcue palette yesterday I perceived the Red indeed darker than the Blue.

I did check again with https://webaim.org/resources/contrastchecker/ and you're right: with the new threshold we get ~4.1-3 for the Red, Blue, Purple which is considered okay for normal text.

Did you try `105` also for the waveforms?
If you find that value working for both Hotcues and waveforms, let me know.

**Comment by Holzhaus** [comment]
> Did you try `105` also for the waveforms?
> If you find that value working for both Hotcues and waveforms, let me know.

Yes, looks good to me.

**Comment by ronso0** [comment]
Thanks!
Are we done here?

**Review by Holzhaus — APPROVED** [review]

**Review by Holzhaus — CHANGES_REQUESTED** [review]
There are some pre-commit issues, please check.

**Comment by ronso0** [comment]
those are about
```diff
-    int getDimBrightThreshold() { return m_dimBrightThreshold; }
+    int getDimBrightThreshold() {
+        return m_dimBrightThreshold;
+    }
```
Either I reformat all surrounding one-liners or we ignore the complaint. I don't care but I'd like to have some consitency in that file.
(IIRC we already discussed that specific spot and agreed to ignore, but I'm not sure)

**Comment by Holzhaus** [comment]
I'm okay with reformatting all of them.

**Comment by ronso0** [comment]
Ready.

File: src/waveform/renderers/waveformwidgetrenderer.h [hunk 9]
```diff
@@ -74,27 +82,54 @@ class WaveformWidgetRenderer {
...
     }
-    double getZoomFactor() const { return m_zoomFactor;}
-    double getGain() const { return m_gain;}
-    int getTrackSamples() const { return m_trackSamples;}
+    double getZoomFactor() const {
+        return m_zoomFactor;
+    }
+    double getGain() const {
+        return m_gain;
+    }
```
Comment (Holzhaus) [inline src/waveform/renderers/waveformwidgetrenderer.h:99]: Not your fault, but shouldn't this be `CSAMPLE_GAIN` instead of `double`? Anyway, fixing this is out of scope for this PR.

**Review by Holzhaus — APPROVED** [review]
LGTM, thank you
